From 8127ff35694a87c2d60a5b4972e1f4176944372c Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 15:22:03 -0400
Subject: [infiniband] ocrdma: Delete AH table if ocrdma_init_hw fails after AH table creation

Message-id: <24924f9ecdbf3c1f639e8e30b68033f768f7e1e1.1411571481.git.dledford@redhat.com>
Patchwork-id: 96652
O-Subject: [Patch RHEL7 28/48] RDMA/ocrdma: Delete AH table if ocrdma_init_hw fails after AH table creation
Bugzilla: 1081067 1088577
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Devesh Sharma <devesh.sharma@emulex.com>

Cleanup the AH table in error path, if HW initialization fails
after AH table creation.

Signed-off-by: Devesh Sharma <devesh.sharma@emulex.com>
Signed-off-by: Selvin Xavier <selvin.xavier@emulex.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit daac96815e969bd70ed5ad21231be2fc5d99506d)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/ocrdma/ocrdma_hw.c b/drivers/infiniband/hw/ocrdma/ocrdma_hw.c
index e6463cb..55308b6 100644
--- a/drivers/infiniband/hw/ocrdma/ocrdma_hw.c
+++ b/drivers/infiniband/hw/ocrdma/ocrdma_hw.c
@@ -1505,6 +1505,7 @@ static void ocrdma_mbx_delete_ah_tbl(struct ocrdma_dev *dev)
  ocrdma_mbx_cmd(dev, (struct ocrdma_mqe *)cmd);
  dma_free_coherent(&pdev->dev, dev->av_tbl.size, dev->av_tbl.va,
      dev->av_tbl.pa);
+ dev->av_tbl.va = NULL;
  dma_free_coherent(&pdev->dev, PAGE_SIZE, dev->av_tbl.pbl.va,
      dev->av_tbl.pbl.pa);
  kfree(cmd);
@@ -2882,13 +2883,15 @@ int ocrdma_init_hw(struct ocrdma_dev *dev)
   goto conf_err;
  status = ocrdma_mbx_get_phy_info(dev);
  if (status)
-  goto conf_err;
+  goto info_attrb_err;
  status = ocrdma_mbx_get_ctrl_attribs(dev);
  if (status)
-  goto conf_err;
+  goto info_attrb_err;
 
  return 0;
 
+info_attrb_err:
+ ocrdma_mbx_delete_ah_tbl(dev);
 conf_err:
  ocrdma_destroy_mq(dev);
 mq_err:
-- 
1.7.1